opm/simulators/wells/WellContainer.hpp

```diff
@@ -0,0 +1,112 @@
+/*
+  Copyright 2021 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_WELL_CONTAINER_HEADER_INCLUDED
+#define OPM_WELL_CONTAINER_HEADER_INCLUDED
+
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace Opm {
+
+template <class T>
```

Comment:
Is this type `T` expected to be "small" in the common case&mdash;e.g., an arithmetic type or an `enum`&mdash;or do you mostly expect to use this container with user-defined/class types that may be expensive to copy?

Reply:
Good question; I guess the plan is that `T` should reach up to *medium sized* objects - e.g. a  vector over all connections in a well. 

But maybe more interesting - this class might very well be short lived; it is introduced in order to facilitate safe and partial refactoring of the `WellState` class - if the `WellContainer`turns out be useful it can live on - otherwise it should be removed/replaced as part of the refactoring.

Reply:
> I guess the plan is that `T` should reach up to _medium sized_ objects - e.g. a vector over all connections in a well.

In that case it might be prudent to change the interfaces that take a `const T&` and copy the value into the internal `data` into taking forwarding references (`T&&`) and using `std::forward<T>(value)` to assign the pertinent `data` values.  That way we preserve the value categories of the arguments which in turn activates move semantics for temporaries and maintain the existing copy semantics for lvalues.

Reply:
OK - sounds smart; I must admit my experience with `std::forward<T>` is limited; but I have updated the PR as I think it should be. Cool if you take another look.